### PR TITLE
Add a tidy check that checks whether the fluent slugs only appear once

### PR DESCRIPTION
**Pull request by mu001999**
As @Nilstrieb said in https://github.com/rust-lang/rust/pull/121828#issuecomment-1972622855:
> Might make sense to have a tidy check that checks whether the fluent slugs only appear once in the source code and lint for that
there's a tidy check already for sorting

We can get the tidy check error:
```
tidy check
tidy error: /path/to/rust/compiler/rustc_const_eval/messages.ftl: message `const_eval_invalid_align` is not used
tidy error: /path/to/rust/compiler/rustc_lint/messages.ftl: message `lint_trivial_untranslatable_diag` is not used
tidy error: /path/to/rust/compiler/rustc_parse/messages.ftl: message `parse_invalid_literal_suffix` is not used
tidy error: /path/to/rust/compiler/rustc_infer/messages.ftl: message `infer_need_type_info_in_coroutine` is not used
tidy error: /path/to/rust/compiler/rustc_passes/messages.ftl: message `passes_expr_not_allowed_in_context` is not used
tidy error: /path/to/rust/compiler/rustc_passes/messages.ftl: message `passes_layout` is not used
tidy error: /path/to/rust/compiler/rustc_parse/messages.ftl: message `parse_not_supported` is not used
```

r? @Nilstrieb

### CODE REVIEW

File: src/tools/tidy/src/fluent_used.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+use std::path::Path;
+
+use regex::Regex;
+
+lazy_static::lazy_static! {
+    static ref MESSAGE: Regex = Regex::new(r#"(?m)^([a-zA-Z0-9_]+)\s*=\s*"#).unwrap();
+    static ref WORD: Regex = Regex::new(r"\w+").unwrap();
+}
+
+fn filter_fluent(path: &Path) -> bool {
```
Comment (Nilstrieb) [inline src/tools/tidy/src/fluent_used.rs:14]: can this be moved into `fluent_alphabetical` to share some of the fluent handling code (and then to two separate things with the list of messages)?

File: src/tools/tidy/src/fluent_used.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+//! Checks that all Flunt messages appear at least twice
```
Comment (fee1-dead) [inline src/tools/tidy/src/fluent_used.rs:1]:
```suggestion
//! Checks that all Fluent messages appear at least twice
```

File: src/tools/tidy/src/fluent_alphabetical.rs [hunk 3]
```diff
@@ -13,11 +14,18 @@ fn filter_fluent(path: &Path) -> bool {
...
+fn check_alphabetic(
+    filename: &str,
+    fluent: &str,
+    bad: &mut bool,
+    msgs: &mut HashMap<String, String>,
+) {
     let mut matches = MESSAGE.captures_iter(fluent).peekable();
     while let Some(m) = matches.next() {
+        let name = m.get(1).unwrap();
+        msgs.insert(name.as_str().to_owned(), filename.to_owned());
```
Comment (Nilstrieb) [inline src/tools/tidy/src/fluent_alphabetical.rs:26]: What if the message exists already? This seems like a convenient place to sanity check that and error in that case.

File: src/tools/tidy/src/fluent_used.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    while let Some(name) = matches.next() {
+        if let Some((name, filename)) = msgs.remove_entry(name.as_str()) {
+            unused_msgs.insert(name, filename);
+        } else {
+            unused_msgs.remove(name.as_str());
+        }
+    }
+}
+
+pub fn check(path: &Path, msgs: &mut HashMap<String, String>, bad: &mut bool) {
```
Comment (Nilstrieb) [inline src/tools/tidy/src/fluent_used.rs:27]:
```suggestion
pub fn check(path: &Path, all_defined_msgs: &mut HashMap<String, String>, bad: &mut bool) {
```
This makes it clearer that we mean *all* the message there were defined here, some of which may be unused.

File: src/tools/tidy/src/fluent_used.rs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    msgs: &mut HashMap<String, String>,
+    unused_msgs: &mut HashMap<String, String>,
+) {
+    let mut matches = WORD.find_iter(contents);
+    while let Some(name) = matches.next() {
+        if let Some((name, filename)) = msgs.remove_entry(name.as_str()) {
+            unused_msgs.insert(name, filename);
+        } else {
+            unused_msgs.remove(name.as_str());
+        }
```
Comment (Nilstrieb) [inline src/tools/tidy/src/fluent_used.rs:23]: This is quite complicated and counterintuitive and it took me a while until I realized the reason this is correct (it sees every message at least once, as it also looks at ftl files, so if it only appears once it's unused) (a comment would have helped^^).

Suggestion for an easier to understand algorithm: When walking the paths, only check Rust files and then have the simple algorithm of starting with a hash map with all messages and try to remove every word. The remaining messages are unused.

File: src/tools/tidy/src/fluent_used.rs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    }
+}
+
+pub fn check(path: &Path, msgs: &mut HashMap<String, String>, bad: &mut bool) {
+    let mut unused_msgs = HashMap::new();
```
Comment (Nilstrieb) [inline src/tools/tidy/src/fluent_used.rs:28]: With my algorithm suggestion above, it makes sense taking `all_defined_msgs` by value and then doing
```rust
let mut unused_msgs = all_defined_msgs;
```
here

**Review by Noratrieb — COMMENTED** [review]
Did a proper review of it, and I think the algorithm can be simplified. After that, happy to merge! I was surprised to see that the performance of it is still really good, not being slower than before.. I guess compute is just so cheap compared to (my quite fast!) disk :)

File: src/tools/tidy/src/fluent_used.rs (hunk 5 above)
Comment (mu001999) [reply src/tools/tidy/src/fluent_used.rs:23]: But some messaged can be used as a part of other messages, like `parse_sugg_change_inner_attr_to_outer` in the following:
```
parse_sugg_change_inner_attr_to_outer = to annotate the {$item}, change the attribute from inner to outer style

parse_inner_attr_not_permitted = an inner attribute is not permitted in this context
    .label_does_not_annotate_this = {parse_label_inner_attr_does_not_annotate_this}
    .sugg_change_inner_to_outer = {parse_sugg_change_inner_attr_to_outer}

parse_inner_attr_not_permitted_after_outer_attr = an inner attribute is not permitted following an outer attribute
    .label_attr = not permitted following an outer attribute
    .label_prev_attr = previous outer attribute
    .label_does_not_annotate_this = {parse_label_inner_attr_does_not_annotate_this}
    .sugg_change_inner_to_outer = {parse_sugg_change_inner_attr_to_outer}

parse_inner_attr_not_permitted_after_outer_doc_comment = an inner attribute is not permitted following an outer doc comment
    .label_attr = not permitted following an outer doc comment
    .label_prev_doc_comment = previous doc comment
    .label_does_not_annotate_this = {parse_label_inner_attr_does_not_annotate_this}
    .sugg_change_inner_to_outer = {parse_sugg_change_inner_attr_to_outer}
```

Comment (mu001999) [reply src/tools/tidy/src/fluent_used.rs:23]: It is unused in Rust files but in Fluent files.

Comment (Nilstrieb) [reply src/tools/tidy/src/fluent_used.rs:23]: Ah... in that case a comment may have to be enough

**Comment by Nilstrieb** [comment]
Awesome, thanks!
@bors r+

**Comment by Nilstrieb** [comment]
it's in the queue, weird.

**Pull request merged**